### DISCUSSION TITLE
Patch submission clarification

### DISCUSSION THREAD

**Discussion by chedge** [Homebrew/discussions#2576] [Tap maintenance and Homebrew development]
I have a fairly simple patch to fix renameutils building on Apple Silicon (due to the removal of the stat64 family), but I'm not exactly sure what the best way to submit it is.

Should it be inline in the formula itself, or submitted as a separate patch to formula-patches first, then referenced in the formula?

The patch in question:

```
diff --git a/src/common/io-utils.c b/src/common/io-utils.c
index f56ae9d..a6e9ce8 100644
--- a/src/common/io-utils.c
+++ b/src/common/io-utils.c
@@ -89,13 +89,8 @@
 bool
 file_exists(const char *file)
 {
-#if HAVE_LSTAT64
-    struct stat64 statbuf;
-    return (lstat64(file, &statbuf) != -1);
-#else
     struct stat statbuf;
     return (stat(file, &statbuf) != -1);
-#endif
 }
```

**Comment by SMillerDev** [comment]
It should be submitted upstream

**Reply by chedge** [reply]
Right, I forgot to mention, the project is currently unmaintained, so there's no real way to submit it upstream, short of forking out a new instance, which seems like a bit of a hammer to swat a fly! The utility itself works fine on modern systems, and seems to be available in current Linux repos, so it doesn't seem like it needs any maintenance in particular, aside from this patch for macOS on Apple Silicon.

**Reply by clason** [reply]
Is this **hard** stance new? I see plenty of formulas in the repo with patches...

(And sometimes patches are needed _only_ because of the specific build environment in homebrew and therefore understandably not material for upstreaming. What about those? Do you then not want such submissions at all?)

**Reply by SMillerDev** [reply]
As far as I know this has always been Homebrew's stance on patching.

See also https://docs.brew.sh/Formula Cookbook

But just because we want all patches upstreamed doesn't mean no formula will ever have patches. It just means they won't have to be patched for long.

**Reply by clason** [reply]
(First, let me make clear that I'm honestly looking for clarification here, not complaining.)

Yes, I am aware of this, but the phrasing was always "should not", not "may not" -- so allowing for reasonable exception. I also don't remember CI hard failing when detecting a `patch` block in a new formula?

**Reply by SMillerDev** [reply]
It's been there since I started contributing as far as I know so that's about 4 years.

And it's always up to a maintainer to judge a submission. But if it's broken before even making it into homebrew, that's a clear indication that it's probably going to give that same maintainer a headache later on.
